src/operation_manager.rs

```diff
@@ -0,0 +1,530 @@
+                                name: format!("Geometry #{} from {}", 1, op_name),
+                                geometry: value.unwrap_geometry().clone(),
+                                var_ident: var_ident.0,
+                                used: true,
+                                geometry_id: None,
+                            };
+                            new_geometry_metadata.insert(var_ident.0, geometry_metadata);
+                            var_idents.insert(var_ident.0);
+                        }
+                    }
+
+                    for (var_ident, value) in value_set.unused_values {
+                        if var_ident.0 + 1 > last_used_var_ident as u64 {
+                            let geometry_metadata = GeometryMetadata {
+                                name: format!("Geometry #{} from {}", 1, op_name),
+                                geometry: value.unwrap_geometry().clone(),
+                                var_ident: var_ident.0,
+                                used: false,
+                                geometry_id: None,
+                            };
+                            new_geometry_metadata.insert(var_ident.0, geometry_metadata);
+                            var_idents.insert(var_ident.0);
+                        }
+                    }
+
+                    // Since ordering matters in case of stacks, var idents need
+                    // to be sorted.
+                    let mut var_idents: Vec<_> = var_idents.iter().collect();
+                    var_idents.sort();
+
+                    // Geometry metadata are pushed onto geometry stacks in order.
+                    // Geometry handler runs for each unused geometry so it can
+                    // be added to scene.
+                    for var_ident in var_idents {
+                        let mut geometry_metadata = new_geometry_metadata
+                            .remove(&var_ident)
+                            .expect("Failed to remove new geometry");
+
+                        if !geometry_metadata.used {
+                            let geometry_id = geometry_handler(&mut geometry_metadata);
+
+                            geometry_metadata.geometry_id = Some(geometry_id);
+                        }
+
+                        self.geometry_metadata.push(geometry_metadata);
+                        self.geometry_stack.push(1);
+                    }
+
+                    self.selected_ops
+                        .last_mut()
+                        .expect("Failed to mutate last operation")
+                        .status = OpStatus::Finished;
+                }
+            }
+        }
+    }
+
+    /// Populates geometry dropdowns with geometries produced in previous
+    /// operations.
+    pub fn prepare_ops(&mut self) {
+        for (i, selected_op) in self.selected_ops.iter_mut().enumerate() {
+            if i == 0 {
+                continue;
+            }
```

Comment:
Why special case 0? If there's no vars, the dropdown will be empty. You later sum the `num_available_geos` anyway.